# Task
Identify and display rows in the Excel file "/content/Copy of cf-products-barcodes (1).xlsx" that contain either duplicate or empty values in the 'fixed-barcode' column, and then summarize these findings.

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## Load Excel File

### Subtask:
Load the Excel file into a pandas DataFrame.


**Reasoning**:
To load the Excel file and display its initial structure, I will import pandas, read the specified Excel file into a DataFrame named `df`, and then display the first few rows of the DataFrame.



In [ ]:
import pandas as pd

# Load the Excel file into a pandas DataFrame
df = pd.read_excel('/content/Copy of cf-products-barcodes (1).xlsx')

# Display the first few rows of the DataFrame
df.head()

product_id            product_title    vendor product_type  \
0  8788552188053  Ombres G Eyeshadow Quad  Guerlain       Luxury   
1  8788552188053  Ombres G Eyeshadow Quad  Guerlain       Luxury   
2  8788552188053  Ombres G Eyeshadow Quad  Guerlain       Luxury   
3  8788552188053  Ombres G Eyeshadow Quad  Guerlain       Luxury   
4  8788552188053  Ombres G Eyeshadow Quad  Guerlain       Luxury   

                                              handle  \
0  https://cf.com.lb/products/ombres-g-eyeshadow-...   
1  https://cf.com.lb/products/ombres-g-eyeshadow-...   
2  https://cf.com.lb/products/ombres-g-eyeshadow-...   
3  https://cf.com.lb/products/ombres-g-eyeshadow-...   
4  https://cf.com.lb/products/ombres-g-eyeshadow-...   

                                         description      variant_id  \
0  Ombres G eyeshadow quads have an exceptional t...  46052464525461   
1  Ombres G eyeshadow quads have an exceptional t...  46052464558229   
2  Ombres G eyeshadow quads have an exceptional t...  46052464459925   
3  Ombres G eyeshadow quads have an exceptional t...  46052464590997   
4  Ombres G eyeshadow quads have an exceptional t...  46726528794773   

        variant_title      sku  price               color size  available  \
0   011 Imperial Moon  I045870  109.0   011 Imperial Moon  NaN       True   
1   214 Exotic Orchid  I045871  109.0   214 Exotic Orchid  NaN       True   
2      258 Wild Nudes  I049472  109.0      258 Wild Nudes  NaN       True   
3  360 Mystic Peacock  I045872  109.0  360 Mystic Peacock  NaN       True   
4    520 Stellar Glow  I056842  115.0    520 Stellar Glow  NaN      False   

   grams                                              image  fixed-barcode  
0      0  https://cdn.shopify.com/s/files/1/0716/5791/45...  3346470436558  
1      0  https://cdn.shopify.com/s/files/1/0716/5791/45...  3346470436510  
2      0  https://cdn.shopify.com/s/files/1/0716/5791/45...  3346470439634  
3      0  https://cdn.shopify.com/s/files/1/0716/5791/45...  3346470436503  
4      0  https://cdn.shopify.com/s/files/1/0716/5791/45...  3346470621909

## Identify Duplicate and Empty Fixed Barcodes

### Subtask:
Identify and filter all rows where the 'fixed-barcode' column has either duplicate values or is empty (NaN/null).


**Reasoning**:
To identify rows with empty or duplicate 'fixed-barcode' values, I will first create a boolean mask for empty values, then another for duplicate values (including first occurrences), combine these masks, and finally filter the DataFrame to create `df_issues`.



In [ ]:
import numpy as np

# 1. Identify rows where the 'fixed-barcode' column is null or empty
# Convert potential empty strings to NaN first to catch all empty values
df['fixed-barcode'] = df['fixed-barcode'].replace('', np.nan)
empty_barcodes = df['fixed-barcode'].isnull()

# 2. Identify rows where the 'fixed-barcode' column contains duplicate values (including the first occurrence)
duplicate_barcodes = df['fixed-barcode'].duplicated(keep=False)

# 3. Combine these two conditions to create a single boolean mask
combined_mask = empty_barcodes | duplicate_barcodes

# 4. Apply this combined mask to filter and select rows that satisfy either condition
df_issues = df[combined_mask]

# Display the first few rows of the issues DataFrame and its shape
print("Rows with empty or duplicate 'fixed-barcode' values:")
print(df_issues.head())
print(f"\nTotal rows with issues: {len(df_issues)}")

Rows with empty or duplicate 'fixed-barcode' values:
       product_id       product_title               vendor product_type  \
16  8850359648405       Impact Parfum       Tommy Hilfiger       Luxury   
47  8984790597781          Cozy Nails      Samoa Cosmetics       Casual   
48  8984790597781          Cozy Nails      Samoa Cosmetics       Casual   
49  8984690229397  Lip Perfection Duo  Maybelline New York       Casual   
50  8984690229397  Lip Perfection Duo  Maybelline New York       Casual   

                                           handle  \
16       https://cf.com.lb/products/impact-parfum   
47          https://cf.com.lb/products/cozy-nails   
48          https://cf.com.lb/products/cozy-nails   
49  https://cf.com.lb/products/lip-perfection-duo   
50  https://cf.com.lb/products/lip-perfection-duo   

                                          description      variant_id  \
16  Impact Parfum  by Tommy Hilfiger is a Oriental...  46248382693525   
47  The Cozy Nails features: - 

**Reasoning**:
The current subtask to identify rows with empty or duplicate barcodes is complete. The next logical step is to summarize these findings as requested by the main task. This includes counting the number of empty barcodes, unique duplicate barcodes, and the total number of rows with issues.



In [ ]:
empty_barcode_count = df_issues['fixed-barcode'].isnull().sum()
duplicate_barcode_count = df_issues['fixed-barcode'].duplicated(keep='first').sum()

# Count the number of unique duplicate barcode values (not including NaN)
unique_duplicate_values = df_issues['fixed-barcode'][df_issues['fixed-barcode'].duplicated(keep=False) & df_issues['fixed-barcode'].notnull()].nunique()

print("--- Summary of Fixed Barcode Issues ---")
print(f"Total rows with issues: {len(df_issues)}")
print(f"Number of rows with empty 'fixed-barcode': {empty_barcode_count}")
print(f"Number of unique duplicate 'fixed-barcode' values: {unique_duplicate_values}")
print("Note: A 'unique duplicate value' refers to a barcode string that appears more than once in the dataset.")


--- Summary of Fixed Barcode Issues ---
Total rows with issues: 985
Number of rows with empty 'fixed-barcode': 744
Number of unique duplicate 'fixed-barcode' values: 96
Note: A 'unique duplicate value' refers to a barcode string that appears more than once in the dataset.


## Display Filtered Rows

### Subtask:
Display the rows that have either duplicate or empty values in the 'fixed-barcode' column, categorizing them for clarity.


### Rows with Empty Barcodes

**Reasoning**:
Now that the markdown heading for empty barcodes is added, I will display the rows from `df_issues` that specifically have empty or null values in the 'fixed-barcode' column, showing only the relevant product identification and barcode columns as requested.



In [ ]:
empty_barcode_rows = df_issues[df_issues['fixed-barcode'].isnull()]
print("Displaying rows with empty 'fixed-barcode' values:")
if not empty_barcode_rows.empty:
    print(empty_barcode_rows[['handle', 'product_title', 'fixed-barcode']].to_markdown(index=False))
else:
    print("No rows found with empty 'fixed-barcode' values.")

Displaying rows with empty 'fixed-barcode' values:
| handle                                                                                                         | product_title                                                                         |   fixed-barcode |
|:---------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------|----------------:|
| https://cf.com.lb/products/lip-perfection-duo                                                                  | Lip Perfection Duo                                                                    |             nan |
| https://cf.com.lb/products/lip-perfection-duo                                                                  | Lip Perfection Duo                                                                    |             nan |
| https://cf.com.lb/products/lip-perfection-duo                  

### Rows with Duplicate Barcodes

**Reasoning**:
Now that the markdown heading for duplicate barcodes is added, I will display the rows from `df_issues` that specifically have duplicate and non-empty 'fixed-barcode' values, showing only the relevant product identification and barcode columns as requested.



In [ ]:
duplicate_non_empty_barcode_rows = df_issues[df_issues['fixed-barcode'].duplicated(keep=False) & df_issues['fixed-barcode'].notnull()]
print("Displaying rows with duplicate 'fixed-barcode' values (excluding empty values):")
if not duplicate_non_empty_barcode_rows.empty:
    print(duplicate_non_empty_barcode_rows[['handle', 'product_title', 'fixed-barcode']].to_markdown(index=False))
else:
    print("No rows found with duplicate 'fixed-barcode' values.")

Displaying rows with duplicate 'fixed-barcode' values (excluding empty values):
| handle                                                                                                                         | product_title                                                                                            |   fixed-barcode |
|:-------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------|----------------:|
| https://cf.com.lb/products/impact-parfum                                                                                       | Impact Parfum                                                                                            |   7640496672119 |
| https://cf.com.lb/products/cozy-nails                                                                                          | Cozy Nails           

In [ ]:
urls2 = set(empty_barcode_rows['handle'].tolist())
urls2

{'https://cf.com.lb/products/4-mini-ooh-la-lips-1-maxi-free',
 'https://cf.com.lb/products/addictive-indulgence-summer-bundle',
 'https://cf.com.lb/products/amore-mio-aware-bare',
 'https://cf.com.lb/products/amore-mio-soft-serve',
 'https://cf.com.lb/products/amore-mio-the-dark',
 'https://cf.com.lb/products/amore-mio-the-nude',
 'https://cf.com.lb/products/amore-mio-the-red',
 'https://cf.com.lb/products/antati-antimicrobial-stainless-steel-spatula',
 'https://cf.com.lb/products/antati-skin-analyzer-device',
 'https://cf.com.lb/products/besame-smoky-waterproof-kajal-eyeliner',
 'https://cf.com.lb/products/burgundy-glam-makeup-bundle',
 'https://cf.com.lb/products/catwalk-browup-browliner',
 'https://cf.com.lb/products/catwalk-colored-inkliner',
 'https://cf.com.lb/products/clin-doeil-matt-eyeshadow',
 'https://cf.com.lb/products/collistar-wrap-refill-4-100ml',
 'https://cf.com.lb/products/dewy-glow-lip-cheek-collection',
 'https://cf.com.lb/products/fast-bright-glow-duo',
 'https://c

In [ ]:
import requests
import re
import json
from typing import Optional

URLS = [
  'https://cf.com.lb/products/4-mini-ooh-la-lips-1-maxi-free',
 'https://cf.com.lb/products/addictive-indulgence-summer-bundle',
 'https://cf.com.lb/products/amore-mio-aware-bare',
 'https://cf.com.lb/products/amore-mio-soft-serve',
 'https://cf.com.lb/products/amore-mio-the-dark',
 'https://cf.com.lb/products/amore-mio-the-nude',
 'https://cf.com.lb/products/amore-mio-the-red',
 'https://cf.com.lb/products/antati-antimicrobial-stainless-steel-spatula',
 'https://cf.com.lb/products/antati-skin-analyzer-device',
 'https://cf.com.lb/products/besame-smoky-waterproof-kajal-eyeliner',
 'https://cf.com.lb/products/burgundy-glam-makeup-bundle',
 'https://cf.com.lb/products/catwalk-browup-browliner',
 'https://cf.com.lb/products/catwalk-colored-inkliner',
 'https://cf.com.lb/products/clin-doeil-matt-eyeshadow',
 'https://cf.com.lb/products/collistar-wrap-refill-4-100ml',
 'https://cf.com.lb/products/dewy-glow-lip-cheek-collection',
 'https://cf.com.lb/products/fast-bright-glow-duo',
 'https://cf.com.lb/products/fast-bright-hydration-radiance-bundle',
 'https://cf.com.lb/products/flawless-finish-lip-conceal-duo',
 'https://cf.com.lb/products/flawless-glow-full-face-essentials-bundle',
 'https://cf.com.lb/products/fluff-it-up-brow-mascara',
 'https://cf.com.lb/products/frenchy-studio-white-nail-art-brush-nail-polish',
 'https://cf.com.lb/products/fresh-face-essentials-kit',
 'https://cf.com.lb/products/gift-card',
 'https://cf.com.lb/products/gloss-lash-elevate-duo',
 'https://cf.com.lb/products/holiday-beauty-inside-kit',
 'https://cf.com.lb/products/hypoallergenic-liquid-eyeshadow',
 'https://cf.com.lb/products/hypoallergenic-longwear-hydrating-balm-foundation',
 'https://cf.com.lb/products/keep-kissing-lipstick',
 'https://cf.com.lb/products/lash-gloss-ultimate-glam-bundle',
 'https://cf.com.lb/products/lip-oil-balm',
 'https://cf.com.lb/products/lip-perfection-duo',
 'https://cf.com.lb/products/liplight-glossy-luminous-lipgloss',
 'https://cf.com.lb/products/lotus-velvet-lipstick',
 'https://cf.com.lb/products/love-your-shape-2in1-line-and-fill-lipliner',
 'https://cf.com.lb/products/love-your-shape-kajal-intense-eyeliner',
 'https://cf.com.lb/products/love-your-shape-pure-browliner',
 'https://cf.com.lb/products/mist-glow-summer-set',
 'https://cf.com.lb/products/narciso-rodriguez-edpi-discovery-set-3x4ml',
 'https://cf.com.lb/products/never-nude-base-coat',
 'https://cf.com.lb/products/never-nude-soft-serve',
 'https://cf.com.lb/products/never-nude-taste-of-winter',
 'https://cf.com.lb/products/never-nude-the-dark',
 'https://cf.com.lb/products/never-nude-the-neutral',
 'https://cf.com.lb/products/never-nude-the-nude',
 'https://cf.com.lb/products/never-nude-the-pink',
 'https://cf.com.lb/products/never-nude-the-red',
 'https://cf.com.lb/products/never-nude-the-white',
 'https://cf.com.lb/products/never-nude-top-coat',
 'https://cf.com.lb/products/never-nude-tropical-breeze',
 'https://cf.com.lb/products/nude-glam-essentials-bundle',
 'https://cf.com.lb/products/ooh-la-lips-matt-liquid-lipstick',
 'https://cf.com.lb/products/power-plus-renewing-haircare-starter-kit',
 'https://cf.com.lb/products/reve-eau-de-parfum',
 'https://cf.com.lb/products/risette-face-powder',
 'https://cf.com.lb/products/shiseido-modern-matte-powder-lipstick',
 'https://cf.com.lb/products/skin-active-fast-bright-vitamin-c-niacinamide-30x-brightening-booster-day-serum-duo',
 'https://cf.com.lb/products/skin-envy-2in1-high-coverage-12hr-concealer',
 'https://cf.com.lb/products/skin-envy-matte-poreless-skin-12hr-wear-foundation',
 'https://cf.com.lb/products/the-summer-indulgence-set',
 'https://cf.com.lb/products/time-fighting-ritual-set',
 'https://cf.com.lb/products/tropical-bloom-cuticle-elixir-oil',
 'https://cf.com.lb/products/vitamin-c-glow-protect-bundle',
 'https://cf.com.lb/products/vitamin-c-micellar-water-facial-brightening-cleanser-and-makeup-remover-duo',
 'https://cf.com.lb/products/vitamin-c-night-glow-bundle',
 'https://cf.com.lb/products/you-broke-my-heart-mono-eyeshadow',
 'https://cf.com.lb/products/zadig-voltaire-this-is-him-shower-gel-75ml'
]

HEADERS = {
    "User-Agent": "Mozilla/5.0 (compatible; ProductScraper/1.0)"
}

SEEDGROW_REGEX = re.compile(
    r'window\.seedgrow_current_product\s*=\s*({.*?})\s*;',
    re.DOTALL
)


def extract_seedgrow_json(html: str) -> Optional[dict]:
    match = SEEDGROW_REGEX.search(html)
    if not match:
        return None
    return json.loads(match.group(1))


results2 = []
failed2 = []

for idx, url in enumerate(URLS, start=1):
    try:
        response = requests.get(url, headers=HEADERS, timeout=15)
        response.raise_for_status()

        product_json = extract_seedgrow_json(response.text)

        if product_json:
            results2.append(product_json)
            print(f"{idx}. OK")
        else:
            failed2.append(url)
            print(f"{idx}. NO_JSON")

    except Exception:
        failed2.append(url)
        print(f"{idx}. ERROR")


# -------------------------
# SAVE TO FILES
# -------------------------
with open("products_with_empty_barcodes.json", "w", encoding="utf-8") as f:
    json.dump(results2, f, ensure_ascii=False, indent=2)

with open("failed2_urls.json", "w", encoding="utf-8") as f:
    json.dump(failed2, f, ensure_ascii=False, indent=2)


print("\n✅ Finished")
print(f"Saved products: {len(results2)}")
print(f"Failed URLs: {len(failed2)}")

1. OK
2. OK
3. OK
4. OK
5. OK
6. OK
7. OK
8. OK
9. OK
10. OK
11. OK
12. OK
13. OK
14. OK
15. OK
16. OK
17. OK
18. OK
19. OK
20. OK
21. OK
22. OK
23. OK
24. OK
25. OK
26. OK
27. OK
28. OK
29. OK
30. OK
31. OK
32. OK
33. OK
34. OK
35. OK
36. OK
37. OK
38. OK
39. OK
40. OK
41. OK
42. OK
43. OK
44. OK
45. OK
46. OK
47. OK
48. OK
49. OK
50. OK
51. OK
52. OK
53. OK
54. OK
55. OK
56. OK
57. OK
58. OK
59. OK
60. OK
61. OK
62. OK
63. OK
64. OK
65. OK
66. OK
67. OK

✅ Finished
Saved products: 67
Failed URLs: 0


## Summary:

### Data Analysis Key Findings

*   **Total Issues Identified**: A total of 985 rows were found to have issues in the 'fixed-barcode' column.
*   **Empty Barcodes**: 744 rows have an empty or null value in the 'fixed-barcode' column, indicating a significant number of products are missing barcode information.
*   **Duplicate Barcode Values**: 96 unique 'fixed-barcode' values are duplicated across different product entries. This means the same barcode is assigned to multiple distinct products.
    *   For example, barcode `7640496672119` is assigned to both 'Impact Parfum' (product\_id 8850359648405) and 'Men's Impact Parfum' (product\_id 8966359122069).
    *   Other notable duplicate barcodes include `5281011237301`, `4059729393876`, `3380810318692`, `4059729348845`, `20714833008`, `5281011177003`, `5281011181000`, and `3348901729208`, each appearing multiple times across different products.

### Insights or Next Steps

*   **Data Quality Improvement**: The identified empty and duplicate barcodes indicate critical data quality issues that need to be addressed to ensure accurate product identification and inventory management.
*   **Root Cause Analysis**: Investigate the source of these data inconsistencies. This could involve reviewing data entry processes, understanding system integrations, or identifying potential data migration errors to prevent future occurrences.


In [ ]:
import json

file_path = '/content/drive/MyDrive/Upwork/mahmoud/daouk_products.json'

# Read the JSON file
with open(file_path, 'r', encoding='utf-8') as f:
    data = json.load(f)

# Display the first product
if data and isinstance(data, list):
    print("First product from the JSON file:")
    print(json.dumps(data[0], indent=2))
elif data and isinstance(data, dict):
    print("JSON file contains a dictionary. Displaying it:")
    print(json.dumps(data, indent=2))
else:
    print("JSON file is empty or in an unexpected format.")

First product from the JSON file:
{
  "id": 10072737513754,
  "title": "Burberry Hero Coffret Eau De Parfum 100ml+ Eau De Parfum 10ml",
  "handle": "burberry-hero-coffret-eau-de-parfum-100ml-eau-de-parfum-10ml",
  "body_html": "<p>Burberry Hero is a bold, sophisticated fragrance that captures the essence of modern masculinity. This exclusive coffret includes a generous 100ml bottle of Eau de Parfum alongside a convenient 10ml travel atomizer, perfect for on-the-go refinement. The composition blends warm spices with woody undertones, creating a distinctive scent that commands attention and leaves a lasting impression. Ideal for the discerning gentleman who appreciates quality craftsmanship and timeless elegance. An exceptional gift set for those who demand excellence in their fragrance collection.</p>",
  "published_at": "2025-12-14T13:11:09+02:00",
  "created_at": "2025-12-14T13:11:08+02:00",
  "updated_at": "2025-12-14T13:12:11+02:00",
  "vendor": "Burberry",
  "product_type": "Perfum

In [ ]:
BASE_URL = "https://daouk.com/products/"

# Ensure 'data' from daouk_products.json is loaded and is a list
if 'data' in globals() and isinstance(data, list):
    daouk_handles = [item['handle'] for item in data if 'handle' in item]
    DAOUK_URLS = [f"{BASE_URL}{handle}" for handle in daouk_handles]
    print(f"Generated {len(DAOUK_URLS)} URLs for daouk.com.")
    # Display the first few generated URLs as a sanity check
    print("First 5 Daouk URLs:")
    for i in range(min(5, len(DAOUK_URLS))):
        print(DAOUK_URLS[i])
else:
    print("Error: 'data' variable not found or not in expected format. Please ensure daouk_products.json was loaded correctly.")
    DAOUK_URLS = []


Generated 4193 URLs for daouk.com.
First 5 Daouk URLs:
https://daouk.com/products/burberry-hero-coffret-eau-de-parfum-100ml-eau-de-parfum-10ml
https://daouk.com/products/dolce-gabbana-the-one-coffret-after-shave-balm-50ml-eau-de-toilette-100ml-shower-gel-50ml
https://daouk.com/products/giorgio-armani-my-way-coffret-eau-de-parfum-90ml-eau-de-parfum-15ml
https://daouk.com/products/scandal-jean-paul-gaultier-coffret-eau-de-parfum-80ml-parfum-30ml
https://daouk.com/products/scandal-absolu-jean-paul-gaultier-parfum-80ml-parfum-10ml-lotion-75ml


In [ ]:
import requests
import re
import json
import html
import os
import time

# -----------------------------
# CONFIG
# -----------------------------
BACKUP_EVERY = 5000
PROGRESS_FILE = "progress2.txt"
BACKUP_PREFIX = "backup2_"
FINAL_FILE = "final_results2.json"
TIMEOUT = 15
DELAY = 0.25  # be polite

# -----------------------------
# FETCH FUNCTION
# -----------------------------
def fetch_product_data(url: str) -> dict:
    headers = {
        "User-Agent": "Mozilla/5.0 (compatible; ProductScraper/1.0)",
        "Accept": "text/html",
    }

    response = requests.get(url, headers=headers, timeout=TIMEOUT)
    response.raise_for_status()

    pattern = re.compile(
        r'productData\s*:\s*(\{.*?\})\s*,\s*productCollections',
        re.DOTALL
    )

    match = pattern.search(response.text)
    if not match:
        raise ValueError("productData not found")

    raw_json = html.unescape(match.group(1))
    return json.loads(raw_json)

# -----------------------------
# LOAD PROGRESS
# -----------------------------
start_index = 0
if os.path.exists(PROGRESS_FILE):
    with open(PROGRESS_FILE, "r") as f:
        start_index = int(f.read().strip())

print(f"Starting from index: {start_index}")

# -----------------------------
# LOAD URL LIST
# -----------------------------
# urls = [...]  # YOUR 4100 URLs HERE
urls = failed_urls_from_daouk  # assume already defined

total = len(urls)
batch_data = []
all_files = []

# -----------------------------
# MAIN LOOP
# -----------------------------
for i in range(start_index, total):
    url = urls[i]
    print(f"Processing .... {url}")

    try:
        product = fetch_product_data(url)
        batch_data.append(product)
    except Exception:
        pass  # silent fail, resume-safe

    # Save progress
    with open(PROGRESS_FILE, "w") as f:
        f.write(str(i + 1))

    # Backup every N items
    if (i + 1) % BACKUP_EVERY == 0 or i == total - 1:
        batch_num = (i + 1) // BACKUP_EVERY + (1 if (i + 1) % BACKUP_EVERY else 0)
        backup_name = f"{BACKUP_PREFIX}{batch_num:03d}.json"

        with open(backup_name, "w", encoding="utf-8") as f:
            json.dump(batch_data, f, indent=2, ensure_ascii=False)

        all_files.append(backup_name)
        batch_data = []

        print(f"Processed {i + 1}/{total}")

    time.sleep(DELAY)

# -----------------------------
# FINAL MERGE
# -----------------------------
final_results2 = []

for file in sorted(f for f in os.listdir() if f.startswith(BACKUP_PREFIX)):
    with open(file, "r", encoding="utf-8") as f:
        final_results2.extend(json.load(f))

with open(FINAL_FILE, "w", encoding="utf-8") as f:
    json.dump(final_results2, f, indent=2, ensure_ascii=False)

print(f"\n✅ DONE — {len(final_results2)} products saved to {FINAL_FILE}")

Starting from index: 0
Processing .... https://daouk.com/products/enzo-professional-titanium-curling-iron-25mm
Processing .... https://daouk.com/products/loreal-hyaluron-expert-serum
Processing .... https://daouk.com/products/ceriotti-gx4-sterilizer
Processing .... https://daouk.com/products/toppik-hair-building-fibers
Processing .... https://daouk.com/products/wella-professionals-color-touch-plus-5-71
Processing .... https://daouk.com/products/wella-professionals-color-touch-plus-6-7
Processing .... https://daouk.com/products/wella-professionals-color-touch-plus-7-3
Processing .... https://daouk.com/products/copy-of-wella-professionals-color-touch-plus-9-16
Processing .... https://daouk.com/products/wella-professionals-color-touch-plus-66-07
Processing .... https://daouk.com/products/wella-professionals-color-touch-plus-9-01
Processing .... https://daouk.com/products/wella-professionals-color-touch-plus-7-7
Processing .... https://daouk.com/products/wella-professionals-color-touch-plu

In [ ]:
file_path = '/content/daouk_part1.json'

# Read the JSON file
with open(file_path, 'r', encoding='utf-8') as f:
    current_results = json.load(f)

In [ ]:
fetched_urls = set()
for product in current_results:
    if 'handle' in product:
        fetched_urls.add(BASE_URL + product['handle'])

failed_urls_from_daouk = [url for url in DAOUK_URLS if url not in fetched_urls]

print(f"Total URLs in DAOUK_URLS: {len(DAOUK_URLS)}")
print(f"Total successfully fetched URLs: {len(fetched_urls)}")
print(f"Number of failed URLs: {len(failed_urls_from_daouk)}")

if failed_urls_from_daouk:
    print("\nFailed URLs:")
    for url in failed_urls_from_daouk:
        print(url)
else:
    print("\nAll URLs from DAOUK_URLS were successfully fetched.")

Total URLs in DAOUK_URLS: 4193
Total successfully fetched URLs: 3533
Number of failed URLs: 660

Failed URLs:
https://daouk.com/products/enzo-professional-titanium-curling-iron-25mm
https://daouk.com/products/loreal-hyaluron-expert-serum
https://daouk.com/products/ceriotti-gx4-sterilizer
https://daouk.com/products/toppik-hair-building-fibers
https://daouk.com/products/wella-professionals-color-touch-plus-5-71
https://daouk.com/products/wella-professionals-color-touch-plus-6-7
https://daouk.com/products/wella-professionals-color-touch-plus-7-3
https://daouk.com/products/copy-of-wella-professionals-color-touch-plus-9-16
https://daouk.com/products/wella-professionals-color-touch-plus-66-07
https://daouk.com/products/wella-professionals-color-touch-plus-9-01
https://daouk.com/products/wella-professionals-color-touch-plus-7-7
https://daouk.com/products/wella-professionals-color-touch-plus-8-81
https://daouk.com/products/wella-professionals-color-touch-plus-7-75
https://daouk.com/products/we

In [10]:
all_products = current_results + final_results2
print(f"Total number of products after combining: {len(all_products)}")

with open("all_products.json", "w", encoding="utf-8") as f:
    json.dump(all_products, f, indent=2, ensure_ascii=False)

print(f"\n✅ DONE — {len(all_products)} products saved to all_products.json")


Total number of products after combining: 4192

✅ DONE — 4192 products saved to all_products.json


In [16]:
all_products[0]

{'id': 10072737513754,
 'title': 'Burberry Hero Coffret Eau De Parfum 100ml+ Eau De Parfum 10ml',
 'handle': 'burberry-hero-coffret-eau-de-parfum-100ml-eau-de-parfum-10ml',
 'description': '<p>Burberry Hero is a bold, sophisticated fragrance that captures the essence of modern masculinity. This exclusive coffret includes a generous 100ml bottle of Eau de Parfum alongside a convenient 10ml travel atomizer, perfect for on-the-go refinement. The composition blends warm spices with woody undertones, creating a distinctive scent that commands attention and leaves a lasting impression. Ideal for the discerning gentleman who appreciates quality craftsmanship and timeless elegance. An exceptional gift set for those who demand excellence in their fragrance collection.</p>',
 'published_at': '2025-12-14T13:11:09+02:00',
 'created_at': '2025-12-14T13:11:08+02:00',
 'vendor': 'Burberry',
 'type': 'Perfume Gift Sets',
 'tags': ['check1'],
 'price': 11000,
 'price_min': 11000,
 'price_max': 11000,
 

In [18]:
import pandas as pd

def flatten_products(products: list[dict]) -> pd.DataFrame:
    rows = []

    for product in products:
        product_id = product.get("id")
        product_title = product.get("title", "")
        vendor = product.get("vendor", "")
        product_type = product.get("type", "")
        handle = BASE_URL + product.get("handle", "")
        description = product.get("description", "")

        # SAFELY resolve default image
        images = product.get("images") or []
        default_image = images[0] if images else None

        # Prepare default values
        option_colors = []
        option_sizes = []

        # Options can be dict OR list
        options = product.get("options", {})

        if isinstance(options, dict):
            option_colors = options.get("color", [])
            option_sizes = options.get("size", [])
        elif isinstance(options, list):
            for opt in options:
                if isinstance(opt, dict): # Added check: Ensure 'opt' is a dictionary
                    name = (opt.get("name") or "").lower()
                    if name == "color":
                        option_colors = opt.get("values", [])
                    if name == "size":
                        option_sizes = opt.get("values", [])

        for variant in product.get("variants", []):
            variant_title = variant.get("title") or product_title

            # Replace default title
            if variant_title.lower() == "default title":
                variant_title = product_title

            # Fallback to product image
            image = variant.get("image") or default_image

            color = variant_title if variant_title in option_colors else None
            size = variant_title if variant_title in option_sizes else None

            rows.append({
                "product_id": product_id,
                "product_title": product_title,
                "vendor": vendor,
                "product_type": product_type,
                "handle": handle,
                "description": description,
                "variant_id": variant.get("id"), # Changed to use 'id' as 'variant_id'
                "variant_title": variant_title,
                "sku": variant.get("sku"),
                "price": variant.get("price") / 100,
                "color": color,
                "size": size,
                "available": variant.get("available"),
                "barcode": variant.get("barcode"),
                "image": image
            })

    return pd.DataFrame(rows)


df = flatten_products(all_products)

# Display as table in Colab (Excel-ready)
df.head(50)

product_id                                      product_title  \
0   10072737513754  Burberry Hero Coffret Eau De Parfum 100ml+ Eau...   
1   10072733286682  Dolce & Gabbana The One Coffret After Shave Ba...   
2   10072732041498  Giorgio Armani My Way Coffret Eau De Parfum 90...   
3   10072731943194  Scandal Jean Paul Gaultier Coffret Eau De Parf...   
4   10072731386138  Scandal Absolu Jean Paul Gaultier Parfum 80ml ...   
5   10071060545818                    Montblanc Men's Legend Gift Set   
6    9962932011290  Narciso Rodriguez Ladies For Her Intense EDP 1...   
7   10070993699098                          MARVEL SPIDER MAN KIT EDT   
8   10070993338650                                Cry Babies Gift Set   
9    8385041924378  Disney Donald Duck Eau De Toilette Spray For K...   
10  10070988685594                           HELLO KITTY EDT 30ML KIT   
11  10070987145498                               MICKEY MOUSE EDT KIT   
12  10070985244954             Paco Rabanne Black Xs L'Exces Gift Set   
13  10070983344410                        Terre D'Hermes EDT Gift Set   
14  10070976233754  Paco Rabanne 1 Million Golden Oud (M) Set Parf...   
15   8877281476890  NARCISO RODRIGUEZ FOR HER EAU DE TOILETTE COFFRET   
16  10070973776154           Jeanne Arthes Js Magnetic Power Gift Set   
17  10070966501658  Britney Spears, Private Show - Eau De Parfum 1...   
18  10070961389850  Jeanne Arthes Rocky Man EDT 100ml + Deodorant ...   
19  10070958375194              Jean Arthes Js Magnetic Sport Coffret   
20  10070955458842                      Andreia Fiber Base Gel Polish   
21  10070955458842                      Andreia Fiber Base Gel Polish   
22  10070955458842                      Andreia Fiber Base Gel Polish   
23  10070955458842                      Andreia Fiber Base Gel Polish   
24  10070955458842                      Andreia Fiber Base Gel Polish   
25  10070955458842                      Andreia Fiber Base Gel Polish   
26  10069889057050           Andreia Square Nail File 100/180 1ANF014   
27  10069887844634          Andreia Diamond Nail File 100/180 1ANF012   
28  10069886959898          Andreia Diamond Nail File 100/100 1ANF011   
29  10069886894362         Andreia Halfmoon Nail File 100/180 1ANF010   
30  10069886009626         Andreia Halfmoon Nail File 180/180 1ANF009   
31  10069885157658         Andreia Halfmoon Nail File 150/150 1ANF008   
32  10069884895514         Andreia Halfmoon Nail File 100/100 1ANF007   
33  10069880733978         Andreia Straight Nail File 100/180 1ANF006   
34  10069879423258         Andreia Straight Nail File 180/180 1ANF005   
35  10069877948698         Andreia Straight Nail File 150/150 1ANF004   
36  10069867036954                 Andreia Straight Nail File 100/100   
37  10069865267482                    Andreia Hybrid Gel Fusion Color   
38  10069865267482                    Andreia Hybrid Gel Fusion Color   
39  10069865267482                    Andreia Hybrid Gel Fusion Color   
40  10069865267482                    Andreia Hybrid Gel Fusion Color   
41  10069865267482                    Andreia Hybrid Gel Fusion Color   
42  10069865267482                    Andreia Hybrid Gel Fusion Color   
43  10069865267482                    Andreia Hybrid Gel Fusion Color   
44  10069865267482                    Andreia Hybrid Gel Fusion Color   
45  10069865267482                    Andreia Hybrid Gel Fusion Color   
46  10069865267482                    Andreia Hybrid Gel Fusion Color   
47  10069865267482                    Andreia Hybrid Gel Fusion Color   
48  10069865267482                    Andreia Hybrid Gel Fusion Color   
49  10069865267482                    Andreia Hybrid Gel Fusion Color   

               vendor       product_type  \
0            Burberry  Perfume Gift Sets   
1     Dolce & Gabbana  Perfume Gift Sets   
2      GIORGIO ARMANI  Perfume Gift Sets   
3             SCANDAL  Perfume Gift Sets   
4             SCANDAL  Perfume Gift Sets   
5          MONT BLAN

In [22]:
# df.to_csv("daouk.com-products.csv")
# df.to_excel("daouk.com-products.xlsx")
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5530 entries, 0 to 5529
Data columns (total 15 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   product_id     5530 non-null   int64  
 1   product_title  5530 non-null   object 
 2   vendor         5530 non-null   object 
 3   product_type   5530 non-null   object 
 4   handle         5530 non-null   object 
 5   description    5530 non-null   object 
 6   variant_id     5530 non-null   int64  
 7   variant_title  5530 non-null   object 
 8   sku            5082 non-null   object 
 9   price          5530 non-null   float64
 10  color          0 non-null      object 
 11  size           0 non-null      object 
 12  available      5530 non-null   bool   
 13  barcode        5069 non-null   object 
 14  image          5519 non-null   object 
dtypes: bool(1), float64(1), int64(2), object(11)
memory usage: 610.4+ KB


In [ ]:
import requests
import re
import json
import html

def extract_shopify_current_product(url: str) -> dict:
    headers = {
        "User-Agent": "Mozilla/5.0",
        "Accept": "text/html",
    }

    res = requests.get(url, headers=headers, timeout=20)
    res.raise_for_status()
    html_text = res.text

    # Find ALL assignments
    pattern = re.compile(
        r'Shopify\.current_product\s*=\s*(\{[\s\S]*?\})\s*;',
        re.MULTILINE
    )

    matches = pattern.findall(html_text)
    if not matches:
        raise ValueError("Shopify.current_product not found")

    # Take the LAST meaningful assignment
    raw_obj = matches[-1]

    # Remove JS spread syntax {...{ JSON }}
    raw_obj = re.sub(r'^\{\s*\.\.\.\s*\{', '{', raw_obj)
    raw_obj = re.sub(r'\}\s*\}$', '}', raw_obj)

    # Decode HTML entities
    raw_obj = html.unescape(raw_obj)

    # Convert to Python dict
    product = json.loads(raw_obj)

    # Sanity check
    if not product.get("id"):
        raise ValueError("Parsed product is empty or invalid")

    return product


# -----------------------
# USAGE
# -----------------------
if __name__ == "__main__":
    url = "https://feel22.com/products/la-roche-posay-12-pure-vitamin-c-anti-aging-face-serum-for-all-skin-types-30ml"

    product_data = extract_shopify_current_product(url)
    print(json.dumps(product_data, indent=2, ensure_ascii=False))
    # Save to file
    # with open("current_product.json", "w", encoding="utf-8") as f:
    #     json.dump(product_data, f, indent=2, ensure_ascii=False)

    # print("✅ Shopify.current_product extracted successfully")


{
  "id": 8528411197578,
  "title": "La Roche Posay 12% Pure Vitamin C Anti-Aging Face Serum For All Skin Types 30ml",
  "handle": "la-roche-posay-12-pure-vitamin-c-anti-aging-face-serum-for-all-skin-types-30ml",
  "description": "<p>A radiance-boosting anti-aging serum that helps reduce fine lines, improve skin texture, and restore a healthy glow.</p>\n<p data-start=\"515\" data-end=\"536\"><strong data-start=\"515\" data-end=\"534\">Key Benefits</strong></p>\n<ul>\n<li data-start=\"515\" data-end=\"536\">Visibly reduces fine lines and early signs of aging.</li>\n<li data-start=\"515\" data-end=\"536\">Brightens skin tone and boosts radiance.</li>\n<li data-start=\"515\" data-end=\"536\">Improves texture, leaving skin smoother and firmer.</li>\n</ul>\n<p data-start=\"696\" data-end=\"739\"><strong data-start=\"696\" data-end=\"737\"> Active Ingredients:</strong></p>\n<ul>\n<li data-start=\"742\" data-end=\"857\">\n<strong data-start=\"742\" data-end=\"781\">12% Pure Vitamin C (Ascorbi